File: Client/BetterMusicPlayer/Behavior/Types/Track.swift
//
//  Track.swift
//  BetterMusicPlayer
//
//  Created by John Butler on 3/27/26.
//

import Foundation;

struct Track {
    public let id: TrackID;
    public let name: String;
    public let duration: Duration;
    
    public let album: Album;
    public let artists: Array<SimplifiedArtist>;
    
    public let discNumber: Int;
    public let trackNumber: Int;
    public let isExplicit: Bool;
    public let isPlayable: Bool;
    public let isLocal: Bool;

    public let externalIDs: ExternalIDs;
    public let externalURLs: ExternalURLs;
    public let fullDetailsURL: URL;
    public let restrictions: Restrictions;
    
    var uri: SpotifyURI {
        return id.uri
    }
    
    init?(json: Dictionary<String, AnyObject>) {
        guard let albumJson = json["album"] as? Dictionary<String, AnyObject>,
              let album = Album(json: albumJson),
              let rawArtists = json["artists"] as? Array<Dictionary<String, AnyObject>>,
              let discNumber = json["disc_number"] as? Int,
              let durationMS = json["duration_ms"] as? Int,
              let explicit = json["is_explicit"] as? Bool,
              let externalIDsJson = json["external_ids"] as? Dictionary<String, AnyObject>,
              let externalIDs = ExternalIDs(json: externalIDsJson),
              let externalURLsJson = json["external_urls"] as? Dictionary<String, AnyObject>,
              let externalURLs = ExternalURLs(json: externalURLsJson),
              let href = json["href"] as? String,
              let url = URL(string: href),
              let id = json["id"] as? String,
              let isPlayable = json["is_playable"] as? Bool,
              let restrictionsJson = json["restrictions"] as? Dictionary<String, AnyObject>,
              let restrictions = Restrictions(json: restrictionsJson),
              let name = json["name"] as? String,
              let trackNumber = json["track_number"] as? Int,
              let rawType = json["type"] as? String,
              let uri = json["uri"] as? String,
              let isLocal = json["is_local"] as? Bool
        else { return nil; }
        
        if rawType != "track" { return nil; }
        
        self.album = album;
        var artists: Array<SimplifiedArtist> = [];
        for rawArtist in rawArtists {
            if let artist = SimplifiedArtist(json: rawArtist) {
                artists.append(artist);
            } else { return nil; }
        }
        self.artists = artists;
        self.discNumber = discNumber;
        self.duration = Duration.milliseconds(durationMS);
        self.isExplicit = explicit;
        self.externalIDs = externalIDs;
        self.externalURLs = externalURLs;
        self.fullDetailsURL = url;
        self.id = TrackID(id);
        if self.id.uri != SpotifyURI(uri) { return nil; }
        self.isPlayable = isPlayable;
        self.restrictions = restrictions;
        self.name = name;
        self.trackNumber = trackNumber;
        self.isLocal = isLocal;
    }
    
}

File: Server/src/core_types/track.rs
use std::time::Duration;

use crate::core_types::SpotifyURI;

use super::{
    ExternalIDs, ExternalURLs, Restrictions, TrackID, album::Album, artist::SimplifiedArtist,
};

#[derive(Debug, Clone, Eq)]
#[allow(dead_code)]
pub struct Track {
    id: TrackID,
    name: String,
    duration: Duration,
    album: Album,
    artists: Vec<SimplifiedArtist>,
    disc_number: u32,
    track_number: u32,
    is_explicit: bool,
    is_playable: bool,
    is_local: bool,
    external_ids: ExternalIDs,
    external_urls: ExternalURLs,
    full_details_url: String,
    restrictions: Restrictions,
}

impl<'de> serde::Deserialize<'de> for Track {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Ok(RawTrack::deserialize(deserializer)?.into())
    }
}
#[derive(Debug, Clone, serde::Deserialize)]
struct RawTrack {
    id: TrackID,
    name: String,
    duration_ms: u64,
    album: Album,
    artists: Vec<SimplifiedArtist>,
    disc_number: u32,
    track_number: u32,
    is_explicit: Option<bool>,
    is_playable: Option<bool>,
    is_local: Option<bool>,
    external_ids: ExternalIDs,
    external_urls: ExternalURLs,
    href: String,
    restrictions: Option<Restrictions>,
}

impl From<RawTrack> for Track {
    fn from(raw_track: RawTrack) -> Self {
        Track {
            id: raw_track.id,
            name: raw_track.name,
            duration: Duration::from_millis(raw_track.duration_ms),
            album: raw_track.album,
            artists: raw_track.artists,
            disc_number: raw_track.disc_number,
            track_number: raw_track.track_number,
            is_explicit: raw_track.is_explicit.unwrap_or(false),
            is_playable: raw_track.is_playable.unwrap_or(false),
            is_local: raw_track.is_local.unwrap_or(false),
            external_ids: raw_track.external_ids,
            external_urls: raw_track.external_urls,
            full_details_url: raw_track.href,
            restrictions: raw_track.restrictions.unwrap_or(Restrictions::None),
        }
    }
}

impl Track {
    #[allow(dead_code)]
    pub fn uri(&self) -> SpotifyURI {
        SpotifyURI::Track(self.id.clone())
    }

    #[allow(dead_code)]
    pub fn id(&self) -> &TrackID {
        &self.id
    }
    pub fn duration(&self) -> &Duration {
        &self.duration
    }
}

impl PartialEq<TrackID> for Track {
    fn eq(&self, other: &TrackID) -> bool {
        &self.id == other
    }
}

impl PartialEq for Track {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}
